### DISCUSSION TITLE
getFiles not returning images

### DISCUSSION THREAD

**Discussion by mikepianka** [excalidraw/excalidraw#4961] [Q&A] [answered]
I am embedding the Excalidraw package in a React app where the user can open and close the Excalidraw scene like a widget. I am saving the scene's state to local storage with a `useEffect` and am re-initializing the scene with the user's sketches when they reopen the Excalidraw widget. This is working fine for the appState and elements but any image files the user adds don't seem to get persisted. The code block looks like this:

```js
  useEffect(() => {
    return () => {
      // save scene state before unmount
      const elements = excalidrawRef.current.getSceneElements();
      const appState = excalidrawRef.current.getAppState();
      const files = excalidrawRef.current.getFiles();

      localStorage.setItem(
        "excalidrawState",
        serializeAsJSON(elements, appState, files, "local")
      );
    };
  }, []);
```

If I put an onChange listener in the scene like this:

```js
      <Excalidraw
        ref={callbackRef}
        initialData={createInitialData()}
        onChange={(elements, state, files) =>
          console.log(elements, state, files)
        }
      />
```

then I can see the added images getting logged to the console. 

![image](https://user-images.githubusercontent.com/47948499/159602483-c4226384-9b4c-4dfc-bdcd-716ec3f20804.png)

So they are in the scene, but the `excalidrawRef.current.getFiles()` call just returns an empty object so the list of files doesn't get persisted to local storage.

![image](https://user-images.githubusercontent.com/47948499/159602584-3c8baae1-7c03-4f11-9b4c-2fcdf3e83dc2.png)

Any ideas? Thanks!

**Comment by dwelle** [comment, marked as answer]
hey @mikepianka — how are you passing the images on initial load? Reading our codebase it seems we're ignoring the `initialData.files`, which is a mistake (we're storing our images separately from localStorage and load the images asynchronously, we we must have forgot to account for a case where the images are passed right into `initialData`).

**Reply by mikepianka** [reply]
So `initialData.files` actually seems to be working fine for me. It's the `getFiles` that wasn't working as expected.

My workaround was to move saving the scene state to storage out of that `useEffect` and into a debounced function in `onChange` instead, like this:

```js
<Excalidraw
  ref={callbackRef}
  initialData={createInitialData()}
  UIOptions={{ canvasActions: { export: false, loadScene: false } }}
  onChange={debounce((elements, state, files) => {
    sessionStorage.setItem(
      "excalidrawState",
      serializeAsJSON(elements, state, files, "local")
    );
    console.debug("saved sketchpad state to storage");
  }, 500)}
/>
```

Then when the widget re-opens, I initialize the scene with `initialData` like this:

```js
function createInitialData() {
  let initialData: IExcalidrawInitialData = {};

  initialData.scrollToContent = true;

  const ss = sessionStorage.getItem("excalidrawState");

  if (ss) {
    const excalidrawState = JSON.parse(ss);
    initialData.appState = excalidrawState.appState;
    initialData.elements = excalidrawState.elements;
    initialData.files = excalidrawState.files;
  } else {
    initialData.appState = {
      viewBackgroundColor: "#edf2ff",
      currentItemStrokeColor: "#c92a2a",
      currentItemStrokeWidth: 4,
      currentItemFontSize: 36,
    };
  }

  return initialData;
}
```

**Reply by dwelle** [reply]
> So `initialData.files` actually seems to be working fine for me.

Never mind, you're right :).

> It's the `getFiles` that wasn't working as expected.

In your original example, you're retrieving the files in an unmount handler — we're clearing the files array in the `componentWillUnmount` (as a manual garbage collection — it technically shouldn't be necessary for us to do, but we've had issues with memory leaks so better be safe in case we leave a dangling pointer for some image object), so by the time your unmount handler is called the object is already emptied.

> My workaround was to move saving the scene state to storage out of that useEffect and into a debounced function in onChange instead

That's the correct way to do this anyway, as it's safer to save user data continuously. If you later switch to saving to an async database such as IndexedDB (which I would recommend doing because LocalStorage has a 5MB limit), you'd likely not be able to save the data in time if user e.g. closed the browser tab.

**Reply by mikepianka** [reply]
Ah yeah I figured you might be deliberately clearing the images on unmount, that makes sense. And you're right, IDB will be the preferred way to go. Thanks for the help 👍
